src/regularize/diffs.jl

```diff
@@ -1,111 +1,197 @@
 #=
 diffs.jl
 finite differences
-2019-03-96 Jeff Fessler, University of Michigan
+2019-03-06 Jeff Fessler, University of Michigan
+2020-06 N-D version by Steven Whitaker
 =#
 
-export diff_map
+export diff_map, diff_forw, diff_adj
 
 using LinearMapsAA: LinearMapAA
-using Test: @test
+using Test: @test, @test_throws
+
+
+# size after apply `diff` along `dim`
+diff_size(N::Dims, dim::Int) = (N[1:dim-1]..., N[dim] - 1, N[dim+1:end]...)
+
+# corresponding length
+diff_length(N::Dims, dim::Int) = prod(diff_size(N, dim))
+
+
+# check argument validity
+# dims should be Int or AbstractVector{Int} or Dims
+# but using duck typing for simplicity
+function diff_check(N::Dims, dims)
+    (length(unique(dims)) != length(dims)) &&
+        throw(ArgumentError("non-unique dims $dims"))
+	(dim -> diff_check(N, dim)).(dims) # check each dim
```

Comment:
I don't know if I've seen this style before. I would have done `map(dim -> diff_check(N, dim), dims)` instead. But either way works.

Reply:
I had never used it before either and decided to try it and it worked.
But `map` is more readable so I've switched it.